pkg/endpoint/policy.go

```diff
@@ -143,79 +150,143 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+	if !forcePolicyCompute && epNextPolicyRevision >= repoRevision && e.desiredPolicy != nil {
 		e.getLogger().WithFields(logrus.Fields{
-			"policyRevision.next": e.nextPolicyRevision,
-			"policyRevision.repo": revision,
-			"policyChanged":       e.nextPolicyRevision > e.policyRevision,
+			"policyRevision.next": epNextPolicyRevision,
+			"policyRevision.repo": repoRevision,
+			"policyChanged":       epNextPolicyRevision > epPolicyRevision,
 		}).Debug("Skipping unnecessary endpoint policy recalculation")
-
-		return nil
+		repo.Mutex.RUnlock()
+		return nil, nil
 	}
 
 	stats.policyCalculation.Start()
-	if e.selectorPolicy == nil {
+	if selectorPolicy == nil {
 		// Upon initial insertion or restore, there's currently no good
 		// trigger point to ensure that the security Identity is
 		// assigned after the endpoint is added to the endpointmanager
 		// (and hence also the identitymanager). In that case, detect
 		// that the selectorPolicy is not set and find it.
-		e.selectorPolicy = repo.GetPolicyCache().Lookup(e.SecurityIdentity)
-		if e.selectorPolicy == nil {
+		//nolint:staticcheck
+		selectorPolicy = repo.GetPolicyCache().Lookup(securityIdentity)
+		//nolint:staticcheck // linter thinks that Lookup cannot return nil, it can.
+		if selectorPolicy == nil {
 			err := fmt.Errorf("no cached selectorPolicy found")
 			e.getLogger().WithError(err).Warning("Failed to regenerate from cached policy")
-			return err
+			repo.Mutex.RUnlock()
+			return nil, err
 		}
 	}
-	// TODO: GH-7515: This should be triggered closer to policy change
-	// handlers, but for now let's just update it here.
-	if err := repo.GetPolicyCache().UpdatePolicy(e.SecurityIdentity); err != nil {
+
+	// UpdatePolicy ensures the SelectorPolicy is fully resolved.
+	// Endpoint lock must not be held!
+	err = repo.GetPolicyCache().UpdatePolicy(securityIdentity)
+	if err != nil {
 		e.getLogger().WithError(err).Warning("Failed to update policy")
-		return err
+		repo.Mutex.RUnlock()
+		return nil, err
 	}
-	calculatedPolicy := e.selectorPolicy.Consume(e)
+	repo.Mutex.RUnlock()
+
+	// Consume converts a SelectorPolicy in to an EndpointPolicy
+	desiredPolicy := selectorPolicy.Consume(e)
 
 	stats.policyCalculation.End(true)
 
-	// This marks the e.desiredPolicy different from the previously realized policy
-	e.desiredPolicy = calculatedPolicy
+	return &policyGenerateResult{
+		revision:         repoRevision,
+		endpointPolicy:   desiredPolicy,
+		selectorPolicy:   selectorPolicy,
+		securityIdentity: securityIdentity,
+	}, nil
+}
+
+// setDesiredPolicy updates the endpoint with the results of a policy calculation.
+//
+// The endpoint write lock must be held, and not released until the desired policy has
+// been pushed in to the policymaps. This is so that we block ApplyPolicyMapChanges, which
+// has the effect of blocking the ipcache from updating the ipcache bpf map. It is required
+// that any accumulated changes are pushed in to the policymap before the ipcache map.
```

Comment:
:ok_hand: 

To expand, I think it would help to mention in the comment what the consequences are if someone were to break that constraint: transient policy drops.

Reply:
done.